torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -1369,6 +1369,41 @@ Tensor sparse_sparse_matmul_backward(
   return _sparse_matrix_mask(b_grad.coalesce(), b.coalesce());
 }
 
+Tensor sparse_coo_constructor_backward(
+    const Tensor& grad,
+    const Tensor& result) {
+  if (grad.device().type() == c10::kCPU) {
+    auto values_grad = result.clone();
+    values_grad._values().fill_(1.);
+    const auto nonzero_values_grad = grad.mul(values_grad);
+    values_grad._values().zero_();
+    values_grad.add_(nonzero_values_grad);
+    return values_grad._values();
+  } else {
+    // add_ modifies values in-place on the CPU, while the CUDA version allocates new values.
+
```

Comment:
@cpuhrsch , what do you think? Is it a bug or a feature?

Reply:
I think this whole function was written in a strange way. Ideally we'd have one code path for CPU and CUDA. But even the CPU code path looks strange and seems to have been a hack to dodge adding more operator coverage. This likely been a few years since someone has touched this, so it might even make sense to consider rewriting this from scratch at least as an exercise to see what's missing.

Reply:
Well, we are still missing fast union kernels. This means that CUDA`s `add` suffers from the very same issue `mul` suffered from, i.e. it is poorly parallelized.

Reply:
Well we can figure out the API and back the needed functions with fragments of this implementation as an intermediate step. At least that'll then increase the impact of the fast union kernel and gives us less mess to worry about.

Reply:
I have come up with a unified smart algorithm that is as fast as our `mul/coalesce` implementation with a `n log n` type of performance complexity. No need to implement anything external, the whole point of this stack is performance and getting rid of `sparse_mask_helper`.